### PR TITLE
solution for issue #32

### PR DESCRIPTION
**Pull request by t-r-student**

### CODE REVIEW

**Comment by sergeylukichev** [comment]
👍

File: src/main/java/de/budget/project/controller/CategoryController.java [hunk 1]
```diff
@@ -28,13 +30,20 @@ public CategoryWebDto createCategory(@RequestBody CategoryWebDto categoryWebDto)
     @GetMapping("/categories/{id}")
     @ResponseBody
     public CategoryWebDto getCategoryById(@PathVariable Long id) {
-        return convertToDto(categoryService.getCategoryById(id));
+        Category category = categoryService.getCategoryById(id);
+        if (category == null){
+            throw new CategoryTypeNotFoundException("Category by id: " + id + " is not found");
+        } else
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/controller/CategoryController.java]: don't need else here, throw wont allow you return anything

File: src/main/java/de/budget/project/controller/TransactionController.java [hunk 2]
```diff
@@ -34,20 +43,31 @@ public Long createTransaction(@RequestBody @Valid TransactionWebRequest transact
     @GetMapping("/transactions/{id}")
     @ResponseBody
     public TransactionWebResponse getTransactionById(@PathVariable("id") Long id) {
-        return convertToWebResponse(transactionService.getTransactionById(id));
+        Transaction transaction = transactionService.getTransactionById(id);
+        if (transaction == null){
+            throw new TransactionNotFoundException("Transaction by id " + id + " is not found");
+        } else
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/controller/TransactionController.java]: `else`

File: src/main/java/de/budget/project/controller/TransactionController.java [hunk 3]
```diff
@@ -34,20 +43,31 @@ public Long createTransaction(@RequestBody @Valid TransactionWebRequest transact
...
     }
 
     @GetMapping("/transactions/user/{id}")
     @ResponseBody
     public List<TransactionDAO> getTransactionsByUserId(@PathVariable("id") Long id) {
-        return transactionService.getTransactionsByUserId(id);
+        List<TransactionDAO> transactionDAOS = transactionService.getTransactionsByUserId(id);
+        if(transactionDAOS.isEmpty()){
+            throw new TransactionNotFoundException("There are no transactions belongs to user by id " + id);
+        } else
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/controller/TransactionController.java]: `else`

File: src/main/java/de/budget/project/error/ErrorResponse.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        this.errorTypeCode = errorTypeCode;
+    }
+
+    public ErrorType getErrorType() {
+        return errorType;
+    }
+
+    public void setErrorType(ErrorType errorType) {
+        this.errorType = errorType;
+    }
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/error/ErrorResponse.java]: Won't lombock replace getters and setters?

File: src/main/java/de/budget/project/exception/exceptions/CategoryNotFoundException.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package de.budget.project.exception.exceptions;
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/exception/exceptions/CategoryNotFoundException.java]: do we really need extra exception class for this? Why `ResourceNotFoundException` is not good enough?

File: src/main/java/de/budget/project/exception/exceptions/CategoryTypeNotFoundException.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package de.budget.project.exception.exceptions;
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/exception/exceptions/CategoryTypeNotFoundException.java]: do we really need extra exception class for this? Why `ResourceNotFoundException` is not good enough?

File: src/main/java/de/budget/project/exception/exceptions/TransactionNotFoundException.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package de.budget.project.exception.exceptions;
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/exception/exceptions/TransactionNotFoundException.java]: do we really need extra exception class for this? Why `ResourceNotFoundException` is not good enough?

File: src/main/java/de/budget/project/model/types/CategoryType.java [hunk 8]
```diff
@@ -56,5 +58,9 @@ public static CategoryType findCategoryTypeName(String name) {
         return categoryType.orElseThrow(() -> new IllegalArgumentException("Name cannot be null"));
     }
 
-
+    @Override
+    @JsonValue
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/model/types/CategoryType.java]: already discussed, we need better solution here

File: src/main/java/de/budget/project/model/types/CurrencyType.java [hunk 9]
```diff
@@ -63,4 +65,10 @@ public static CurrencyType findCurrencyByName(String name) {
                 .findFirst();
         return currency.orElseThrow(() -> new IllegalArgumentException("Name cannot be null"));
     }
+
+    @Override
+    @JsonValue
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/model/types/CurrencyType.java]: already discussed, we need better solution here

File: src/main/java/de/budget/project/repository/CategoryRepository.java [hunk 10]
```diff
@@ -11,4 +13,9 @@
     Category getCategoryById(Long id);
 
     List<Category> getAllByCategoryType(CategoryType categoryType);
+
+    @Query("SELECT new de.budget.project.model.dao.CategoryDAO" +
+            "(c.id, c.name, c.categoryType)" +
+            "FROM Category c WHERE c.categoryType = 1 OR c.categoryType = 2")
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/repository/CategoryRepository.java]: `FROM Category c WHERE c.categoryType = 1 OR c.categoryType = 2`
is it possible to have category entry without type? For what reason do we need to specify here which types should it have especially when we have only 2 types and both of them are requested. Could we use just '*' instead?

File: src/main/java/de/budget/project/services/InitialDataService.java [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package de.budget.project.services;
```
Comment (MichelRogov) [inline src/main/java/de/budget/project/services/InitialDataService.java]: don't think we need extra service for this, everything can be done in controller. If we would have any logic, then yes.

File: src/main/java/de/budget/project/controller/CategoryController.java (hunk 1 above)
Comment (t-r-student) [reply src/main/java/de/budget/project/controller/CategoryController.java]: Resolved

File: src/main/java/de/budget/project/controller/TransactionController.java (hunk 2 above)
Comment (t-r-student) [reply src/main/java/de/budget/project/controller/TransactionController.java]: Resolved

File: src/main/java/de/budget/project/controller/TransactionController.java (hunk 3 above)
Comment (t-r-student) [reply src/main/java/de/budget/project/controller/TransactionController.java]: Resolved

File: src/main/java/de/budget/project/error/ErrorResponse.java (hunk 4 above)
Comment (t-r-student) [reply src/main/java/de/budget/project/error/ErrorResponse.java]: Resolved by adding @Data

File: src/main/java/de/budget/project/repository/CategoryRepository.java (hunk 10 above)
Comment (t-r-student) [reply src/main/java/de/budget/project/repository/CategoryRepository.java]: Resolved: "SELECT new de.budget.project.model.dao.CategoryDAO" +
            "(c.id, c.name, c.categoryType)" +
            "FROM Category c"

**Pull request merged**